### PR TITLE
Make REST parameters schema-susceptible

### LINKED ISSUE #14060: Internal Server Error 500 returned when filtering with a field that doesn't exist in the table schema
### Required System information

- Node.js version: v16.16.0
- NPM version: 8.11.0
- Strapi version: 4.3.2
- Database: Mysql
- Operating system: Debian GNU/Linux 10 (buster)

### Describe the bug

I was trying to filter the data with (filters[active][$eq]=true)  but the field active doesn't exist in the table, the expected response in this case is 200 with empty data but I get a 500 Internal Server Error
 
### Steps to reproduce the behavior

1. Try to filter with a field that doesn't exist in the table 

### Expected behavior

I think the expected response, in this case, is status Code 200 with empty data Or a response ignoring the filter.

### Screenshots

Screen Shot 2022-08-10 at 12 47 17

Thank you

### PR DESCRIPTION
**Pull request by WalkingPizza**
### What does it do?

Alters `convertSortQueryParams`, `convertFiltersQueryParams` and `convertFieldsQueryParams` so that all attributes not present in the schema are automatically removed from the query.

### Why is it needed?

At the moment, trying to sort by, filter on or select a field that does not exist in the schema throws an Internal Server Error.

### How to test it?

Run queries that have a `fields`, `sort` or `filters` parameter that contain a field that does not exist on the schema and notice how the server automatically ignores those parameters.

This, of course, also applies to `fields`, `sort` or `filters` nested inside a `populate` object.

### Related issue(s)/PR(s)

Should fix #14060

### CODE REVIEW

**Comment by alexandrebodin** [comment]
Hey @WalkingPizza I think we might have to check this in the DB.

The idea to not be restrictive at that level was to allow filtering on manually created columns in db for example. I have to admit I don't think this is necessary but if we want to keep this we would need to handle the check at a lower level

**Comment by WalkingPizza** [comment]
My bad, I didn't know that was possible!

Should I try to implement this at the DB level?
I would have to find a way to list the columns in a specific table, have the query builder convert sort/filter/fields parameters to their column name and then perform the checks.

Definitely trickier, but I am sure it can be done.

**Comment by alexandrebodin** [comment]
So we largely do all this already and check if the query is on an existing column or throw. I think the thing missing is more the dedicated error handle that would return a 400 but I'm skeptical in terms approach to even have to do that here...

**Comment by WalkingPizza** [comment]
Let me know if you find a better way of doing this, I'd be more than happy to modify the PR.

The only reason why I am more keen on ignoring than throwing is that if we throw every time people try to query a non-existing field, we are technically leaking parts of our data structure, and that might be a bit of a security concern.

But no biggie, I've definitely done worse in some of my project 😂

**Comment by alexandrebodin** [comment]
That's why you usually just throw a 500 to avoid giving any details. My pbl with ignoring things is that the DX isn't great either as if they have a typo they will just get some data that might not be right and not know why.

I guess I would add a warning in the code and leave the 500 happen so the user can look at the logs and know what happened instead of leaking this through the API 🤔

**Comment by Marc-Roig** [comment]
Hello! I was reading the PR and the thread. Wanted to see if we could close issue https://github.com/strapi/strapi/issues/14060.

In this thread, The conclusion was, to not leak the db structure, when filtering (sorting, …) with a non existent field, we should return a 500 status error + a warning in the console.

Now, in the console, you see the raw sql error. Do you think this should be enough?
image

Looking at each database dialect, there is a transformError . But there is no way to see if you are asking for a non existent column (at least for sqlite).

**Comment by alexandrebodin** [comment]
The more I think about it and I think we could review  @WalkingPizza's PR. As long as the DB layer allows for more fields but from the API you can only filters on existing fields in the schema that should be a good default behavior.

File: packages/core/utils/lib/convert-query-params.js [hunk 1]
```diff
@@ -300,7 +336,7 @@ const convertAndSanitizeFilters = (filters, schema) => {
 
   // Here, `key` can either be an operator or an attribute name
   for (const [key, value] of Object.entries(filters)) {
-    const attribute = get(key, schema.attributes);
+    const attribute = key === 'id' ? { type: 'integer' } : get(key, schema.attributes);
```
Comment (Marc-Roig) [inline packages/core/utils/lib/convert-query-params.js:339]: why is this change needed?

Comment (Marc-Roig) [reply packages/core/utils/lib/convert-query-params.js:339]: If it is that way, my question would be (more for @alexandrebodin ?), why the `id` is not in the schema attributes.

**Review by Marc-Roig — APPROVED** [review]
LGTM! :) 
Everything is working on my machine.

File: packages/core/utils/lib/convert-query-params.js (hunk 1 above)
Comment (WalkingPizza) [reply packages/core/utils/lib/convert-query-params.js:339]: Yep, it's for the filter by ID functionality!
And the ID is not in the attributes because [the spec](https://jsonapi.org/format/) says it has to be a top-level field.

Comment (Marc-Roig) [reply packages/core/utils/lib/convert-query-params.js:339]: oh! makes sense! Thank you @WalkingPizza  😃

**Comment by WalkingPizza** [comment]
Hey @alexandrebodin, will this be merged eventually?
No pressure, just wondering!

**Comment by alexandrebodin** [comment]
Hey @WalkingPizza, Can you update your PR and fix the tests 1st ?

**Comment by alexandrebodin** [comment]
@Marc-Roig  do you mind handling it to help @WalkingPizza  out? we are pretty much ready to merge it

**Comment by Marc-Roig** [comment]
> @Marc-Roig do you mind handling it to help @WalkingPizza out? we are pretty much ready to merge it

Sure, I will fix tests and update the branch

**Comment by innerdvations** [comment]
Thanks for shedding light on this topic so long ago, and I'm glad to say it's finally been responsible for some movement!

Closing this since it's stale and we've now fixed `filters` with a slightly different method. If you'd like to update the remaining `sort` and `fields` based off the current `main` branch, feel free to go ahead with a new PR, otherwise we will probably be adding those in the coming month or so, and I'll try to make sure you get credit in the release notes :)
